### ISSUE TITLE
Build using Py_LIMITED_API and then build wheels with the stable ABI abi3 tag

### ISSUE THREAD

**Issue by mattip** [pytorch/pytorch#52134]
## 🚀 Feature
Pytorch could build python c-extension modules with `#define Py_LIMITED_API`, which would allow using the stable ABI `abi3` tag, so that one c-extension module (and wheel) would work on multiple CPython versions.
## Motivation
Limit the number of wheels needed for each release

## Pitch

PyTorch wheels are very large, straining the PyPI infrastructure as well as requiring significant resources on a user' machine. Providing a more universal wheel would ease that burden

## Additional context

The [1.7.1 release](https://pypi.org/project/torch/#files) has 12 wheels: 3 for each python version 3.6, 3.7, 3.8, 3.9. If sucessful, this enhancement would reduce that to 3 wheels.

**Comment by ezyang** [comment]
It is an interesting proposal. We'd have to find out:
1. How much of the non `Py_LIMITED_API` we are currently using, and
2. How much performance cost it would be to stop using it.
I know that @robieta was pretty psyched about FASTCALL which is recent versions of Python only, and we would have to stop using that, at the very least

**Comment by robieta** [comment]
Yeah, it would be a shame to give up `METH_FASTCALL`, as the preliminary measurements I did indicate that it can reduce our arg parsing overhead by high single to low double digit percentages.

**Comment by mattip** [comment]
OK. I guess we will have to wait for until someone implements [`PyCCall_FastCall`](https://www.python.org/dev/peps/pep-0579/#meth-fastcall-is-private-and-undocumented) and it becomes part of the stable API. There was an attempt in [PEP 580](https://www.python.org/dev/peps/pep-0580/) that was rejected. Shame, it could have saved some overhead.

**Comment by fzimmermann89** [comment]
Is https://docs.python.org/3/c-api/call.html#vectorcall (part of the stable abi since 3.12) enough?

**Comment by rgommers** [comment]
Pybind11 does not support the limited C API, and PyTorch relies heavily on it. So this seems completely inactionable, unless there's a major effort to change from Pybind11 to Nanobind.

**Comment by albanD** [comment]
tbh pybind11 is not even the worst offender. Both Dynamo (relying on the frame hook API) and our Tensor object binding (relying on subtype dealloc logic and gc behavior) heavily rely on implementation details of CPython (we even have quite a bit of their C code).
